File: release/policy_check/challenge.py
"""Canonical signer proof challenge parser."""

from datetime import timedelta
from typing import Any
import re

from .common import FINGERPRINT, GIT_ID, parse_datetime

FIELD_NAMES = [
    "repository",
    "primary_fingerprint",
    "role",
    "authority_revision",
    "nonce",
    "issued_at",
    "expires_at",
]


def challenge_valid(case: dict[str, Any]) -> bool:
    if set(case) != {
        "id", "challenge", "expected_authority_revision", "observed_at",
        "observed_primary_fingerprint", "verifier_result", "expected",
    }:
        return False
    challenge = case["challenge"]
    if not isinstance(challenge, str) or not challenge.isascii() or not challenge.endswith("\n"):
        return False
    if "\r" in challenge or challenge.endswith("\n\n"):
        return False
    lines = challenge[:-1].split("\n")
    if len(lines) != 8 or lines[0] != "prikk-release-signer-proof-v1":
        return False
    fields: dict[str, str] = {}
    for expected_name, line in zip(FIELD_NAMES, lines[1:], strict=True):
        if not line.startswith(expected_name + "="):
            return False
        fields[expected_name] = line[len(expected_name) + 1 :]
    if fields["repository"] != "https://github.com/nabbisen/prikk":
        return False
    if not FINGERPRINT.fullmatch(fields["primary_fingerprint"]):
        return False
    if fields["role"] != "official-release" or not GIT_ID.fullmatch(fields["authority_revision"]):
        return False
    if not re.fullmatch(r"[0-9a-f]{64}", fields["nonce"]):
        return False
    try:
        issued = parse_datetime(fields["issued_at"])
        expires = parse_datetime(fields["expires_at"])
        observed = parse_datetime(case["observed_at"])
    except (TypeError, ValueError):
        return False
    if not issued < expires or expires - issued > timedelta(hours=24):
        return False
    if issued - observed > timedelta(minutes=5) or observed >= expires:
        return False
    return (
        case["expected_authority_revision"] == fields["authority_revision"]
        and case["observed_primary_fingerprint"] == fields["primary_fingerprint"]
        and isinstance(case["verifier_result"], str)
        and bool(case["verifier_result"])
    )

File: tools/release-policy/src/policy/challenge.rs
use std::collections::BTreeMap;

use serde_json::Value;

use super::signer::fingerprint;
use crate::time::parse_utc_second;

const FIELD_NAMES: [&str; 7] = [
    "repository",
    "primary_fingerprint",
    "role",
    "authority_revision",
    "nonce",
    "issued_at",
    "expires_at",
];

pub(super) fn validate(context: &Value, challenge: &[u8]) -> Option<&'static str> {
    let Some(object) = context.as_object() else {
        return Some("challenge-grammar-or-binding");
    };
    if object.len() != 5
        || ![
            "id",
            "observed_at",
            "observed_primary_fingerprint",
            "verifier_result",
            "expected_authority_revision",
        ]
        .iter()
        .all(|name| object.contains_key(*name))
    {
        return Some("challenge-grammar-or-binding");
    }
    let Ok(text) = std::str::from_utf8(challenge) else {
        return Some("challenge-grammar-or-binding");
    };
    if !text.is_ascii() || !text.ends_with('\n') || text.contains('\r') || text.ends_with("\n\n") {
        return Some("challenge-grammar-or-binding");
    }
    let lines: Vec<&str> = text[..text.len() - 1].split('\n').collect();
    if lines.len() != 8 || lines.first() != Some(&"prikk-release-signer-proof-v1") {
        return Some("challenge-grammar-or-binding");
    }
    let mut fields = BTreeMap::new();
    for (name, line) in FIELD_NAMES.iter().zip(lines.iter().skip(1)) {
        let prefix = format!("{name}=");
        let Some(value) = line.strip_prefix(&prefix) else {
            return Some("challenge-grammar-or-binding");
        };
        fields.insert(*name, value);
    }
    if fields.get("repository") != Some(&"https://github.com/prikk-vcs/prikk")
        || fields.get("role") != Some(&"official-release")
        || !fields
            .get("primary_fingerprint")
            .is_some_and(|value| fingerprint(value))
        || !fields
            .get("authority_revision")
            .is_some_and(|value| git_id(value))
        || !fields.get("nonce").is_some_and(|value| {
            value.len() == 64
                && value
                    .bytes()
                    .all(|byte| byte.is_ascii_hexdigit() && !byte.is_ascii_uppercase())
        })
    {
        return Some("challenge-grammar-or-binding");
    }
    let issued = fields
        .get("issued_at")
        .and_then(|value| parse_utc_second(value));
    let expires = fields
        .get("expires_at")
        .and_then(|value| parse_utc_second(value));
    let observed = object
        .get("observed_at")
        .and_then(Value::as_str)
        .and_then(parse_utc_second);
    let (Some(issued), Some(expires), Some(observed)) = (issued, expires, observed) else {
        return Some("challenge-time-window");
    };
    if issued >= expires
        || expires - issued > 24 * 60 * 60
        || issued - observed > 5 * 60
        || observed >= expires
    {
        return Some("challenge-time-window");
    }
    if object
        .get("expected_authority_revision")
        .and_then(Value::as_str)
        != fields.get("authority_revision").copied()
        || object
            .get("observed_primary_fingerprint")
            .and_then(Value::as_str)
            != fields.get("primary_fingerprint").copied()
        || object
            .get("verifier_result")
            .and_then(Value::as_str)
            .is_none_or(str::is_empty)
    {
        return Some("challenge-grammar-or-binding");
    }
    None
}

fn git_id(value: &str) -> bool {
    matches!(value.len(), 40 | 64)
        && value
            .bytes()
            .all(|byte| byte.is_ascii_hexdigit() && !byte.is_ascii_uppercase())
}

#[cfg(test)]
#[path = "challenge/tests.rs"]
mod tests;
